app/src/components/code/JSONEditor.tsx

```diff
@@ -28,32 +46,79 @@ export type JSONEditorProps = Omit<
 
 export function JSONEditor(props: JSONEditorProps) {
   const { theme } = useTheme();
+  const { jsonSchema, ...restProps } = props;
+  const [isVisible, setIsVisible] = useState(false);
+  const [isInitialized, setIsInitialized] = useState(false);
+  const containerRef = useRef<HTMLDivElement>(null);
   const codeMirrorTheme = theme === "light" ? githubLight : nord;
   const extensions = useMemo(() => {
     const baseExtensions = [
       json(),
       EditorView.lineWrapping,
       linter(jsonParseLinter()),
     ];
-    if (props.jsonSchema) {
+    if (jsonSchema) {
       baseExtensions.push(
         linter(jsonSchemaLinter(), { needsRefresh: handleRefresh }),
         jsonLanguage.data.of({
           autocomplete: jsonCompletion(),
         }),
         hoverTooltip(jsonSchemaHover()),
-        stateExtensions(props.jsonSchema)
+        stateExtensions(jsonSchema)
       );
     }
     return baseExtensions;
-  }, [props.jsonSchema]);
+  }, [jsonSchema]);
+
+  /**
+   * The two useEffect hooks below are used to initialize the JSON editor.
+   * This is necessary because code mirror needs to calculate its dimensions to initialize properly.
+   * When it is rendered outside of the viewport, the dimensions may not always be calculated correctly,
+   * resulting in the editor being invisible or cut off when it is scrolled into view.
+   * Below we use a combination of an intersection observer and a delay to ensure that the editor is initialized correctly.
+   * For a related issue @see https://github.com/codemirror/dev/issues/1076
+   */
+  useEffect(() => {
+    const observer = new IntersectionObserver(([entry]) => {
+      setIsVisible(entry.isIntersecting);
+    });
+
+    if (containerRef.current) {
+      observer.observe(containerRef.current);
+    }
+    const current = containerRef.current;
+
+    return () => {
+      if (current) {
+        observer.unobserve(current);
+      }
+    };
+  }, []);
+
+  useEffect(() => {
+    if (isVisible && !isInitialized) {
+      // Small delay to ensure DOM is ready
+      const timer = setTimeout(() => {
+        setIsInitialized(true);
+      }, JSON_EDITOR_INITIALIZATION_DELAY);
+      return () => clearTimeout(timer);
+    }
+  }, [isInitialized, isVisible]);
+
   return (
-    <CodeMirror
-      value={props.value}
-      extensions={extensions}
-      editable
-      theme={codeMirrorTheme}
-      {...props}
-    />
+    <div
+      ref={containerRef}
+      css={css`
+        min-height: ${!isInitialized ? JSON_EDITOR_MIN_HEIGHT : "auto"};
```

Comment:
this is to make it so this container fits the size of the editor after it's been initialized